### PR TITLE
add support for validation loss

### PR DESCRIPTION
**Pull request by xlr8harder**
Hi, I needed a validation loss because I have no good intuitions with audio models.  I've only connected it to LoRA sft finetuning, because I only have SFT data and my laptop can only handle LoRA, but it should be pretty easy to port to the others.  I'm glad to port it to the others too, if you think it's useful, I just won't be able to test them thoroughly.

### CODE REVIEW

File: csm_mlx/cli/finetune/lora_finetune.py [hunk 1]
```diff
@@ -206,6 +207,40 @@ def sft_finetune(
...
+        Optional[Path],
+        typer.Option(
+            "--val-data-path",
+            help="Path to JSON dataset file with text/audio pairs",
+            exists=True,
+            file_okay=True,
+            dir_okay=False,
+            readable=True,
+        )
+    ] = "",
```
Comment (senstella) [inline csm_mlx/cli/finetune/lora_finetune.py]: For type safety purposes, how about using `= None` as a default?

File: csm_mlx/cli/finetune/lora_finetune.py [hunk 2]
```diff
@@ -309,34 +347,59 @@ def sft_finetune(
     )
     print(f"Loaded {len(dataset)} samples")
 
+    val_dataset = []
```
Comment (senstella) [inline csm_mlx/cli/finetune/lora_finetune.py]: Also for type safety purposes, `None` as a default would be preferrable, as that matches `val_dataset: CSMDataset | None = None` signature.

File: csm_mlx/finetune/trainer.py [hunk 3]
```diff
@@ -317,6 +321,54 @@ def compute_loss(
...
+            total_val_loss += batch_loss * batch_size_actual
+            total_samples += batch_size_actual
+            pbar_val.set_postfix({"val_loss": f"{batch_loss:.4f}"})
+
+            self.model.train()
+
+            if total_samples == 0:
+                return float('inf')
+            avg_val_loss = total_val_loss / total_samples
+            return avg_val_loss
```
Comment (senstella) [inline csm_mlx/finetune/trainer.py]: I think it needs to explicitly cast mx.array to float, like: `float(avg_val_loss)`.

File: csm_mlx/finetune/trainer.py [hunk 4]
```diff
@@ -379,7 +431,8 @@ def _step(batch: Dict[str, mx.array]):  # type: ignore
         return float(loss)
 
     def train(
-        self, dataset: CSMDataset, batch_size: int, epochs: int, shuffle: bool = True
+        self, dataset: CSMDataset, batch_size: int, epochs: int,
+        val_dataset: CSMDataset | None = None, shuffle: bool = True
```
Comment (senstella) [inline csm_mlx/finetune/trainer.py]: I think it can break some methods like: `return super().train(dataset, batch_size, epochs, shuffle)` such as in `KTOTrainer`. How about 

```py
self, dataset: CSMDataset, batch_size: int, epochs: int, shuffle: bool = True, *, val_dataset: CSMDataset | None = None
```

as a signature?

File: csm_mlx/cli/finetune/lora_finetune.py [hunk 5]
```diff
@@ -297,6 +332,9 @@ def sft_finetune(
             gradient_checkpointing=gradient_checkpointing,
             ckpt_freq=ckpt_freq,
             log_freq=log_freq,
+            val_freq=val_freq,
```
Comment (senstella) [inline csm_mlx/cli/finetune/lora_finetune.py]: CLI's `val_freq` and `val_batch_size` can be `None` or `int` while the TrainArg's those two cannot be `None`. How about doing something like `= val_freq or 0`?

**Comment by senstella** [comment]
Thank you for the nice PR! Adding validation loss is really helpful - I was planning to implement it but hadn't gotten around to it yet, so I really appreciate you taking care of it. It's such a valuable addition.

I've added some comments, mostly just typing fixes and minor adjustments. Also, please note that the logic for full fine-tuning should be the same as with LoRA, if you might want to add the command for `full` too. Again, I really appreciate the PR, thanks!

File: csm_mlx/finetune/trainer.py (hunk 4 above)
Comment (xlr8harder) [reply csm_mlx/finetune/trainer.py]: I've never worked with typing much before, so welcome feedback here.

I did this, but it also looks like I needed to update the type signature on the train() calls on the classes derived from CSMTrainer.  I updated them to match and they also pass along the val_dataset to the super().train() call if they receive one, but it wasn't clear to me at first glance if they would actually work.  Any idea?

Should I raise an error instead if it's passed to KTOTrainer / DPOTrainer, or do you think it might work as is?

**Comment by xlr8harder** [comment]
I'm glad to add these changes to full once you're happy with the implementation, but I don't think I can test a full finetune on my laptop.

File: csm_mlx/finetune/trainer.py (hunk 4 above)
Comment (senstella) [reply csm_mlx/finetune/trainer.py]: Since both `DPOTrainer` and `KTOTrainer` require additional parameters to be attached on the batch, I think it would be nice to raise an error when it's passed on them for now - I'll implement the custom `evaluate` for those trainers and remove the errors later on once it's merged!

**Comment by senstella** [comment]
Thank you! And everything looks nice to me! So, I don't think there will be any issues even if we add the evaluating functionality to the full-finetuning CLI. Only difference between two CLI is the part of weight being freezed, so `.eval()` and `.train()` switching would work smoothly. No testing needed!

**Comment by xlr8harder** [comment]
sorry for the delay, I probably won't be able to finish this for another day or two.

I will add the code to raise the error in the DPOTrainer and KTOTrainer, and copy the implementation over to the non-LoRA trainer, as well.

**Comment by xlr8harder** [comment]
Oh, I thought there was a separate trainer for full finetune, but it looks like I only needed to add the command line options?  If so, then this should be done.

I've also added the check to KPOTrainer and DPOTrainer to raise an error if a val_dataset is passed in, as discussed.

**Comment by senstella** [comment]
Thank you so much! Everything looks perfect now with those options and checks. I'm happy to merge this in now. Great work and thanks again for your contribution!
